### DISCUSSION TITLE
C++ Exceptions

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#109] [Ideas]
To make working with C++ exceptions easier I added tools to translate general C++ exceptions (`RuntimeError`) to specific types. You can see the implementation [here](https://github.com/gswifort/PyRx/blob/50eeafd5f2bf89deaacabe86546358abd73339aa/pyrx/exceptions.py).
Usage looks like this:

```py
text = Db.Text()
text_id = db.addToModelspace(text)
with cast_to_arx_exception:
    Db.Text(text_id, Db.OpenMode.kForRead)  # raises pyrx.exceptions.eWasOpenForWrite
```
I wonder if I should stick to this path or if you have a better idea..? Maybe you want to implement something similar in C++?

**Comment by CEXT-Dan** [comment]
This is what I am currently using for exceptions, it appears to be limited to RuntimeError and a message
https://boostorg.github.io/python/doc/html/reference/high_level_components/boost_python_exception_translato.html

**Reply by CEXT-Dan** [reply]
That's pretty cool though. 
for support requests, I'll want to have the C++ exception message, Acad::ErrorStatus, function, Line, File.

**Reply by gswifort** [reply]
Yes, in the example I only showed the error type, but the message is also passed so the message is complete.

**Comment by gswifort** [comment]
See if this will be helpful.

https://stackoverflow.com/questions/2261858/boostpython-export-custom-exception

**Comment by CEXT-Dan** [comment]
Okay, great idea, seems very useful!
I’ll have a look at this, though I have a couple other tasks in front

**Comment by CEXT-Dan** [comment]
I was able to get the sample from stack overflow working (the one by David Faure towards the end). I don’t know if I like it as it’s kind of a monkey patch.

The sample attaches an object ‘cause’, which I renamed ctx, short for context to the built-in exception.
There’s no IntelliSense since it’s a patch, maybe this can be done manually

ctx has properties 
```Python
err.ctx.fullmessage
err.ctx.message
err.ctx.code
```

example:
```Python

from pyrx_imp import Ap, Db, Ed, Ge, Gi, Gs, Rx

def PyRxCmd_edoit1():
    try:
        db = Db.curDb()
        text = Db.Text()
        text_id = db.addToModelspace(text)
        Db.Text(text_id, Db.OpenMode.kForRead)
    except Db.ErrorStatusException as err:
        match err.ctx.code:
            case Db.ErrorStatus.eWasOpenForWrite:
                print("caught eWasOpenForWrite")
            case Db.ErrorStatus.eOnLockedLayer:
                print("caught eOnLockedLayer")
            case _:
                print("caught {}".format(err.ctx.code))
    except Exception as err:
        print(err)
        
def PyRxCmd_edoit2():
    try:
        db = Db.curDb()
        text = Db.Text()
        text_id = db.addToModelspace(text)
        Db.Text(text_id, Db.OpenMode.kForRead)
    except Db.ErrorStatusException as err:
        match err.ctx.message:
            case 'eWasOpenForWrite':
                print("caught eWasOpenForWrite")
            case 'eOnLockedLayer':
                print("caught eOnLockedLayer")
            case _:
                print("caught {}".format(err.ctx.code))
    except Exception as err:
        print(err)

def PyRxCmd_doit1():
    try:
        db = Db.curDb()
        text = Db.Text()
        text_id = db.addToModelspace(text)
        Db.Text(text_id, Db.OpenMode.kForRead)
    except Db.ErrorStatusException as err:
        print(1,err.ctx.fullmessage)
    except Exception as err:
        print(4,err)
        
def PyRxCmd_doit4():
    try:
        db = Db.curDb()
        text = Db.Text()
        text_id = db.addToModelspace(text)
        Db.Text(text_id, Db.OpenMode.kForRead)
    except Exception as err:
        print(4,err)
```

**Reply by CEXT-Dan** [reply]
I guess I need to make a build, so people can kick it around

**Reply by gswifort** [reply]
Looks good.
Can't you add `fullmessage`, `message` and `code` directly to `ErrorStatusExeption`?

**Reply by CEXT-Dan** [reply]
That’s what the first example attempts to do that, but apparently it does not work with Python 3
Also, I think I would have to maintain a static state, at that point it would be the same as adding global last error function

‘ctx’ is an actual object that gets garbage collected so there’s no memory leak
Maybe there’s a better name for ctx?

**Reply by gswifort** [reply]
No, the `ctx` name is ok.

**Reply by CEXT-Dan** [reply]
Another, maybe better approach is to assign a tuple to the exception with function 
PyErr_SetObject 

example
```Python
def PyRxCmd_doit5():
    try:
        db = Db.curDb()
        text = Db.Text()
        text_id = db.addToModelspace(text)
        Db.Text(text_id, Db.OpenMode.kForRead)
    except Db.ErrorStatusException as err:
        es, sm, fm = err.args
        
        print(es)
        print(sm)
        print(fm)
        
    except Exception as err:
        print(err)
```

I think I could attach a dict, this would prevent issues with ordering

**Reply by gswifort** [reply]
Assigning a `dict` to `args` can be confusing, maybe `namedtuple`?

**Reply by CEXT-Dan** [reply]
actually, I figured out how to add the methods directly

```Python
def PyRxCmd_doit1():
    try:
        db = Db.curDb()
        text = Db.Text()
        text_id = db.addToModelspace(text)
        Db.Text(text_id, Db.OpenMode.kForRead)
    except Db.ErrorStatusException as err:
        print(err.code)
        print(err.message)
        print(err.fullmessage)
    except Exception as err:
        print(err)
```
